### PR TITLE
Support KSP in BindsMethodValidator

### PR DESCRIPTION
**Pull request by IlyaGulya**
Part of the issue #751

### CODE REVIEW

**Comment by IlyaGulya** [comment]
@ZacSweers 
This one is almost done.
There's three tests that are failing because they require dagger compilation:
https://github.com/square/anvil/blob/540ce1095b3b0bf153ca2897b6ed378424dbb048/compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt#L272
https://github.com/square/anvil/blob/540ce1095b3b0bf153ca2897b6ed378424dbb048/compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt#L318
https://github.com/square/anvil/blob/540ce1095b3b0bf153ca2897b6ed378424dbb048/compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt#L373

As far as I can see, dagger compilation with ksp is not yet supported in tests (according to https://github.com/square/anvil/blob/540ce1095b3b0bf153ca2897b6ed378424dbb048/compiler/src/test/java/com/squareup/anvil/compiler/TestUtils.kt#L214).
Do I need to do something with these three tests?

**Comment by IlyaGulya** [comment]
Oh, I forgot to write a test for this issue: #750 
Will do that tomorrow 🙁

**Comment by ZacSweers** [comment]
For those three tests, what you wanna do is enable the KSP validator on the first compilation and then just use the regular dagger processing on the subsequent compilation. Hoping that could work?

Sidenote - there are some related changes I've made in #830 to support Dagger's new behavior in 2.48, but you're welcome to cherry-pick those into this PR since they sorta go together and I can update my PR after this is in

**Comment by ZacSweers** [comment]
eh actually the more I think about it, the more I think it'll be smoother for me to just update that PR and get this one through first. I can manage any conflicts. Just know that it's there :)

**Comment by IlyaGulya** [comment]
> eh actually the more I think about it, the more I think it'll be smoother for me to just update that PR and get this one through first. I can manage any conflicts. Just know that it's there :)

Ok, now I will fix those three tests and add a test for #750, after that I will wait for your PR to be merged 🙂

**Comment by IlyaGulya** [comment]
@ZacSweers regarding the #750 - I've added test which is correctly passes in KSP mode.
However, I've spent about 4 hours already trying to fix it in Embedded mode and I think it will require another 4-8 hours to get any success =)
Will Embedded mode be replaced by KSP mode or it will stay and I should fix this behavior?

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/ClassReference.kt
Comment (IlyaGulya) [inline compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/ClassReference.kt:230]: There's only one place that I'm a bit worried about. I'm not sure if there will be performance impact.
Should I cache class descriptor instead of resolving it every time?

**Comment by ZacSweers** [comment]
if it's not too difficult, do you think it's possible to split out the fix for #750 to a separate PR first? Might make this easier to review

**Comment by IlyaGulya** [comment]
> if it's not too difficult, do you think it's possible to split out the fix for #750 to a separate PR first? Might make this easier to review

Sure!

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/ClassReference.kt
Comment (IlyaGulya) [reply compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/ClassReference.kt:230]: Extracted into https://github.com/square/anvil/pull/833

**Comment by IlyaGulya** [comment]
@ZacSweers I've extracted #750 fix to https://github.com/square/anvil/pull/833

**Comment by ZacSweers** [comment]
Let's start reviewing this once https://github.com/square/anvil/pull/833 is in and this branch is updated. Suspect there will be a decent amount of overlap :)

**Comment by ZacSweers** [comment]
@IlyaGulya ping me when you've pulled latest main and we can start review here!

**Comment by IlyaGulya** [comment]
@ZacSweers done!
I've also removed the receiver parameter validation because extension functions are no longer supported.

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt
Comment (ZacSweers) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt:357]: nice

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt
Comment (ZacSweers) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt:384]: Can you add a small comment explaining why we need this for posterity?

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ksp/KspUtil.kt
Comment (ZacSweers) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/ksp/KspUtil.kt:118]: Nice

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/BindsMethodValidator.kt
Comment (ZacSweers) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/BindsMethodValidator.kt:139]: Can we use KSP's built-in assignability check? `KSType.isAssignableFrom(...)`

**Review by ZacSweers — COMMENTED** [review]
Nice work

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt
Comment (IlyaGulya) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/BindsMethodValidatorTest.kt:384]: I'm not sure how to explain it in a comment.
I've added this flag to fix tests according to your suggestion: https://github.com/square/anvil/pull/831#issuecomment-1869815259

**Review by ZacSweers — APPROVED** [review]

**Comment by ZacSweers** [comment]
Haven't forgotten about this, we just had to revert the embedded impl temporarily due to issues :/. Will revisit once that one is in again, unless you want to rebase this one

**Comment by IlyaGulya** [comment]
@ZacSweers you mean this revert? https://github.com/square/anvil/commit/5a04d6086c5a2925f13a052a691ceae3e7edbce9
I haven't got time to fix that issue yet, unfortunately.
But I can rebase this PR if it will help, for sure 🙂

**Comment by IlyaGulya** [comment]
@ZacSweers done 🙂

**Review by ZacSweers — APPROVED** [review]
Sorry for taking so long on this, appreciate your patience!

**Pull request merged**
